Speedup triangulation of edges

Czaki opened the pull request:
# References and relevant issues

# Description

Similar to #7256 this PR is a speedup of face triangulation by avoiding a lot of reallocation of arrays.

After long investigation, I found that using best way to speed this up is by avoiding the allocation of additional arrays by using numba.  

For shapes that do not have bevels:

On main:

![obraz](https://github.com/user-attachments/assets/b6775d94-a88b-4a99-959f-f4b6f7f768aa)

In this PR:

![obraz](https://github.com/user-attachments/assets/9a3a302c-2eb1-4ecd-8c38-a07617512ede)

PR with cache:

![obraz](https://github.com/user-attachments/assets/6d7f00a8-4d55-4bb2-9f49-955bc746e01f)



As you may see, the time of `triangulate_edge` is reduced from 44s to 17s. Half of it is numba compilation.

With sharp edges that require bevel join, the speedup is even bigger:

Main:

![obraz](https://github.com/user-attachments/assets/e2a55f13-6ba7-4f1d-9ff9-56fa46fdf49c)

PR:
![obraz](https://github.com/user-attachments/assets/6279350e-1e50-441e-a538-4f01b5420d40)


PR with cache:

![obraz](https://github.com/user-attachments/assets/3c992928-b63c-42d7-be6d-d0ec7ec620f5)


This PR also fixes the bevel join 

On main:

![obraz](https://github.com/user-attachments/assets/9a24bc31-9777-4d60-81d5-ff6cc1b0d1a2)


With PR:

![obraz](https://github.com/user-attachments/assets/b167a4d5-9bce-442c-80c6-d5a0c90e5f8d)


In this PR I have added `examples/dev/triangle_edge.py` with a set of helper functions to visualize triangulation. These are helpful for debugging of triangulation problems. 

# Algorithm explanation

Here I add some explanation to the code implementation of this PR. 

## Calculation of miter vector <a href="#user-content-miter" id="miter">#</a>

This is a screenshot of miter join:

![obraz](https://github.com/user-attachments/assets/08d2dd78-bb02-4da8-baa5-7b0596e1c3c7)

And let's assume markings as on image:

![obraz](https://github.com/user-attachments/assets/d0cec9ff-3786-4edc-805a-d45e58f33aab)


The $\vec{BA}$ (blue) is half of the normal vector from vertex $n$ to $n+1$. The $\vec{AC}$ is minus half of the normal vector from vertex $n-1$ to $n$. The vector $\vec{AD}$ (green) is orthogonal to line $BA$ of length $0.5$ 

Based on these vectors, we could easily calculate $\vec{BC} = \vec{BA} + \vec{AC}$. However, the length of $|GC|$ may be different from 1. We would like to get the vector $\vec{BE}$ 

We will use here trigonometry and [Intercept theorem](https://en.wikipedia.org/wiki/Intercept_theorem)

We know that $\frac{|GC|}{|AC|} = \sin{\alpha}$ and $|AC| = \frac{1}{2}$ so $|GC| = \frac{\sin{\alpha}}{2}$ 


From intercept theorem, we know that $\frac{|BC|}{|GC|} = \frac{|BE|}{|FE|}$. The length of 
$|FE| = \frac{1}{2}$. So $|BE| = |BC| \times \frac{1}{\sin{\alpha}}$. That mens that $\vec{BE} = \vec{BC} \times \frac{1}{\sin{\alpha}}$

## Bevel limit <a href="#user-content-bevel-limit" id="bevel-limit">#</a>

In general, we use miter join, however, if the angle is sharp, it may result in strange artifacts (see #6706). So if the join will be too long (The length of $\vec{BA}$ vector from above), we would like to change to a bevel join.

So we need to calculate the length of $\vec{BE}$. We would like to avoid calculation of square root in every turn of the loop. 

From Pythagoras theorem we know that $|BE|^2 = |EF|^2 + |BF|^2 = \frac{1}{2}^2 + |BF|^2$

We also know that $|GA| = \frac{\cos{\alpha}}{2}$ so $|BG| = \frac{1 - \cos{\alpha}}{2}$. From miter vector calculation and intercept theorem we know that $|BF| = \frac{|BG|}{\sin{\alpha}} = \frac{1 - \cos{\alpha}}{2\sin{\alpha}}$ 

So we know the length of the vector 

$$|BE|^2 = \frac{1}{2}^2 + \left(\frac{1 - \cos{\alpha}}{2\sin{\alpha}}\right)^2 = \frac{\sin^2{\alpha} + 1^2 - 2\cos{\alpha} + \cos^2{\alpha}}{4\sin^2{\alpha}}
= \frac{2 - 2 \cos{\alpha}}{4\sin^2{\alpha}}
$$

Using trigonometric identity

$$|BE|^2 = \frac{1}{2} \cdot \frac{1 - \cos{\alpha}}{1 - \cos^2{\alpha}} = 
\frac{1}{2} \cdot \frac{1 - \cos{\alpha}}{(1 - \cos{\alpha})(1 + \cos{\alpha})} = 
\frac{1}{2} \cdot \frac{1}{1+\cos{\alpha}}$$ 

So if we would like to check condition $|BE| > limit$ we could use the following equation: 

$$ \frac{1}{2} \cdot \frac{1}{1+\cos{\alpha}} > {limit}^2$$ 

So:

$$ \frac{1}{2 \cdot {limit}^2} - 1 > \cos{\alpha} $$ 

It means that if $\cos\alpha$ is below a given threshold, we should use bevel join.

## Too long vector in bevel join <a href="#user-content-bevel-cut" id="bevel-cut">#</a>

As I showed above, this PR is fixing the bevel join by fixing the position of inner edge of the bevel join. You may see an example in this image (see the longest yellow vector):

![obraz](https://github.com/user-attachments/assets/58e13821-a1b6-4dd0-a49b-ade97edf9e73)

The schematic of generating long edge:

![obraz](https://github.com/user-attachments/assets/329b2b61-3c9d-4050-9623-7053c1f46248)

As users may use shapes with different length edges, we cannot use a hard-coded limit for the length of the offset vector that creates the problems pointed on above image. 

However, when calculating the normal vectors, we need to calculate the edge width, so we may save them for later usage and limit the length of such problematic offset vector to length of shortest adjusted edge. 

In the Bevel limit section, we have an explicit formula for miter offset length. However, it contains a square root. And calculation of  a square root is expensive on current CPU, so we would like to avoid it.  

However, the described problem happens only when $\measuredangle BAC$ is close to $180\degree$, but in such a situation, the $|\vec{BA} + \vec{AC}|$ is close to 1 (as resulted vector is close to diameter).  

So we could approximate the length of vector $\vec{BC}$ as 1. Then the estimated length of vector $\vec{BE}$ is $\frac{1}{\sin{\alpha}}$ 

So if $\frac{1}{\sin{\alpha}}$ is bigger than the length of half of any adjusted edges, we could calculate a new scaling factor with value 

$$\pm 0.5 \cdot\min(vec1\textunderscore{}len, vec2\textunderscore{}len))$$

# Additional notes

## Artifacts

There are still artifacts in rendering 

![obraz](https://github.com/user-attachments/assets/416df018-f293-425c-bd38-916ba8137934)

Czaki committed:
Speedup triangulation of edges

Czaki committed:
Merge branch 'main' into faster_faces

Czaki committed:
add faster edge triangulation

Czaki committed:
add basic tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix straight line case

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix polygon test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix triangle orientation

Czaki committed:
improve bevel join

Czaki committed:
fix enforce bevel

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix type annotation

Czaki committed:
skjip test when missing numba

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
remove duplicated points in poligon

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix shapes dtype

Czaki committed:
set of fixes to keep lasso tool working...

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
improve test coverage

Czaki committed:
Merge remote-tracking branch 'upstream/main' into faster_faces

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

melonora commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,318 @@
+        Nx2 or Nx3 array of central coordinates of deduplicated path
+    """
+    if len(path) <= 2:
+        # part of ugly hack to keep lasso tools working
```
is there an open issue on this that can be referred to?

Czaki replied:
No. In general, problem comes from adding two points in some places (as removing points duplication is causing that more polygons pass `is_covex`), Also new code for triangulation of edges assumes no duplicates on path.

Czaki replied:
Maybe I missed something, but if there will be no remarks changing my understanding I will open such Issue.

willingc replied:
@Czaki Let's open an issue and refer to it in the comment. :sunny:

jni replied:
@Czaki can you explain what the issue is exactly? My interpretation:

> when the lasso tool "drops" a new point in a polygon being drawn, momentarily, the last to points in the polygon being drawn are identical. This happens also at the very beginning of drawing, when the first point is added to the polygon and the second point (which follows the mouse/pen tip) is also added immediately. Because of this, if we only have two points, we ignore it if they are duplicates.

Czaki committed:
update docstrings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix cos limit calculation

Czaki committed:
update docstring with references to PR description

Czaki committed:
improve variable naming

Czaki committed:
make example interactive

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
move loop to separate function

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
refactor code for better readability

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,480 @@
+    return path_
+
+
+@njit(cache=True)
```
I have tried to remove this decorator, but it doubles execution time. 

With decorator:
![obraz](https://github.com/user-attachments/assets/79782822-c7ee-4dbc-93a3-9695fdafcfdd)

without a decorator:

![obraz](https://github.com/user-attachments/assets/74a8f5d2-d23a-40f2-9c6d-7f5b0a03b9c5)

Czaki committed:
add docstring

Czaki committed:
Merge branch 'main' into faster_faces

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+from __future__ import annotations
```
Adding a 1 line module docstring at top of file would be helpful for future contributors/maintainers.
Perhaps:

```suggestion
"""Triangulation utilities"""
from __future__ import annotations
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    normals : np.ndarray
+        Nx2 array of normal vectors of the path
+    closed : bool
+        Bool if shape edge is a closed path or not.
```
```suggestion
        True if shape edge is a closed path.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        bevel join
+    bevel : bool
+        Bool which if True causes a bevel join to always be used.
+        If False a bevel join will only be used when the miter limit is exceeded
```
```suggestion
        If True, a bevel join is always to be used.
        If False, a bevel join will only be used when the miter limit is exceeded.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+                normals[i - 1, 0] * normals[i, 0]
+                + normals[i - 1, 1] * normals[i, 1]
+            )
+            if cos_limit > cos_angle:
```
```suggestion
            if cos_angle < cos_limit:
```
Subtle change but flipping the check to logically read: If calculated cos_angle (which changes in the else block) is less than cos_limit (which doesn't change in this function).

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+                normals[-2, 0] * normals[-1, 0]
+                + normals[-2, 1] * normals[-1, 1]
+            )
+            if cos_limit > cos_angle:
```
```suggestion
            if cos_angle < cos_limit:
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+            cos_angle = (
+                normals[-1, 0] * normals[0, 0] + normals[-1, 1] * normals[0, 1]
+            )
+            if cos_limit > cos_angle:
```
```suggestion
            if cos_angle < cos_limit:
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        triangles of the triangulation
+    vertex : np.ndarray
+        The vertex of the path for which the centers,
+         offsets and triangles art calculated
```
```suggestion
        offsets and triangles art calculated
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    centers[j + 1] = vertex
+    cos_angle = vec1[0] * vec2[0] + vec1[1] * vec2[1]
+    sin_angle = vec1[0] * vec2[1] - vec1[1] * vec2[0]
+    if sin_angle == 0:
```
```suggestion

    if sin_angle == 0:
```

Add newline to create a more readable block.

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        mitter = np.array([vec1[1], -vec1[0]], dtype=np.float32) * 0.5
+    else:
+        scale_factor = 1 / sin_angle
+        if bevel or cos_limit > cos_angle:
```
```suggestion
        if bevel or cos_angle < cos_limit:
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        Nx2 array of normal vectors of the path
+    closed : bool
+        Bool if shape edge is a closed path or not.
+    cos_limit : float
```
Consider renaming this to `miter_limit` for better readability.

Czaki replied:
Are you sure that comparing `miter_limit` with `cos_angle` will be more readable? 

`cos_angle < miter_limit`

maybe 

`cos_angle < cos_miter_limit`

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    -----
+    we use cos_limit instead of maximum miter length
+    for performance reasons.
+    This is an equivalent check, see note in generate_2D_edge_meshes
```
```suggestion
    We use a miter limit instead of maximum miter length
    for performance reasons.
    This approach provides an equivalent check.
    See note in generate_2D_edge_meshes
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    centers : np.ndarray
+        Mx2 array central coordinates of path triangles.
+    offsets : np.ndarray
+        Mx2 array of the offsets to the central coordinates that need to
```
```suggestion
        Mx2 array of the offsets to the central coordinates. Offsets need to
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+            # For performance reasons here, the mitter length is estimated
+            # by the inverse of the sin of the angle between the two vectors.
+            # See https://github.com/napari/napari/pull/7268#user-content-bevel-cut
+            elapsed_len = scale_factor
```
`elapsed_len` is a bit confusing as a variable name. Is there an alternative name that better describes? Perhaps `scaled_length` or `scaled_len`.

Czaki replied:
My idea for the name of this variable was `estimated_len`. I do not know why I use `elapsed` instead of `estimated`

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        # We use here the Intercept theorem for calculating the mitter length
+        # More details in PR description:
+        # https://github.com/napari/napari/pull/7268#user-content-miter
+        mitter = (vec1 - vec2) * 0.5 * scale_factor
```
```suggestion
        mitter = (vec1 - vec2) * 0.5 * scale_factor
 
```
Add new line for better readability.

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+
+        triangles[j] = [j, j + 1, j + 2]
+
+        return 3  # added 3 triangles because of bevel
```
```suggestion
        triangles[j] = [j, j + 1, j + 2]
        return 3  # bevel join added 3 triangles
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    offsets[j + 1] = -mitter
+    triangles[j] = [j, j + 1, j + 2]
+    triangles[j + 1] = [j + 1, j + 2, j + 3]
+    return 2  # added 2 triangles
```
```suggestion
    return 2  # miter join added 2 triangles
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+) -> None:
+    """Fix the orientation of the triangles.
+
+    For checking if a point is inside a triangle.
```
```suggestion
    The orientation is useful to check if a point is inside a triangle.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    centers : np.ndarray
+        Mx2 array central coordinates of path triangles.
+    offsets : np.ndarray
+        Mx2 array of the offsets to the central coordinates that need to
```
```suggestion
        Mx2 array of the offsets to the central coordinates. Offsets need to
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+def _normal_vec_and_length(
+    path: np.ndarray, closed: bool
+) -> tuple[np.ndarray, np.ndarray]:
+    """Calculate the normal vector and length of the vector.
```
```suggestion
    """Calculate the normal vector and its length.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    path : np.ndarray
+        Mx2 array representing path to calculate the normal vector and length.
+        Assumes that there is no point repetition in the path.
+        (no two consecutive points are the same)
```
```suggestion
        In other words, assume no two consecutive points are the same.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        triangles[0] = [0, 1, 2]
+        triangles[1] = [1, 2, 3]
+        j = 2
+    for i in range(1, len(normals) - 1):
```
```suggestion

    for i in range(1, len(normals) - 1):
```

Add new line to separate the else block.

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+
+@njit(cache=True, inline='always')
+def _cut_end_if_repetition(path: np.ndarray) -> np.ndarray:
+    """Cut the last point of the path if it is the same as the second last point.
```
```suggestion
    """Cut the last point of the path if it is the same as the second to last point.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    return path_
+
+
+@njit(cache=True)
```
```suggestion
# Note: removing this decorator will double execution time.
@njit(cache=True)
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    limit: float = 3.0,
+    bevel: bool = False,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Determines the triangulation of a path in 2D. The resulting `offsets`
```
```suggestion
    """Determines the triangulation of a path in 2D.
    
    The resulting `offsets`
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    `centers` to generate the vertices of the triangles for the triangulation,
+    i.e. `vertices = centers + width*offsets`. Using the `centers` and
+    `offsets` representation thus allows for the computed triangulation to be
+    independent of the line width.
```
```suggestion
    i.e. `vertices = centers + width*offsets`. By using the `centers` and
    `offsets` representation, the computed triangulation can be
    independent of the line width.
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+
+    cos_limit = 1 / (2 * (limit / 2) ** 2) - 1.0
+    # why cos_limit uis calculated this way is explained in the note in
+    # https://github.com/napari/napari/pull/7268#user-content-bevel-limit
```
```suggestion
    # why cos_limit is calculated this way is explained in the note in
    # https://github.com/napari/napari/pull/7268#user-content-bevel-limit
    cos_limit = 1 / (2 * (limit / 2) ** 2) - 1.0
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+
+    # We need to fix triangle orientation, as our code for checking
+    # if points is in triangle is not robust to orientation
+    _fix_triangle_orientation(triangles, centers, offsets)
```
```suggestion
    # We fix triangle orientation to improve checks for
    # whether points are in triangle.
    # Without the fix, the code is not robust to orientation.
    _fix_triangle_orientation(triangles, centers, offsets)
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+        # should be len(path) - 1 but we need to allow
+        # duplication of the last point to keep the lasse tool working
+        if np.all(path[i] == path[i + 1]):
+            dup_count += 1
```
```suggestion
    dup_count = 0
    # We would would like to use len(path) - 1 as the range.
    # To keep the lasso tool working, we need to allow
    # duplication of the last point.
    # If the lasso tool is refactored, update to use the preferred range.
    for i in range(len(path) - 2):
        if np.all(path[i] == path[i + 1]):
            dup_count += 1
```

willingc commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,510 @@
+    new_path = np.empty((target_len, path.shape[1]), dtype=np.float32)
+    index = 0
+
+    new_path[0] = path[0]
```
```suggestion
    target_len = len(path) - dup_count
    new_path = np.empty((target_len, path.shape[1]), dtype=np.float32)
    new_path[0] = path[0]
    index = 0
```

willingc reviewed (commented):
@Czaki I've done a partial review of this PR.

Most of my suggestions are for improved readability and maintainability of the code by others on the team.

Nice work!

willingc commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,259 @@
+    First vector is the normal vector scaled by 1/2
+    Second vector is the normal vector of previous point scaled by -1/2
+    """
+    vec2 = normal_vector.copy()
```
Any particular reason that `vec2` and `vec3` are used instead of `vec1` and `vec2`?

Czaki replied:
No. Only result of splitting functions of smaller one.

willingc commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,259 @@
+    vec5[:, 0] += vec2[:, 1]
+    vec5[:, 1, 0] = -vec2[:, 1, 1]
+    vec5[:, 1, 1] = vec2[:, 1, 0]
+    return vec5
```
Since this is an example, perhaps:
```suggestion
    vec1 = normal_vector.copy()
    vec1[:, 1] *= 0.5
    orthogonal_vector = vec1.copy()
    
    orthogonal_vector[:, 0] += vec1[:, 1]
    orthogonal_vector[:, 1, 0] = -vec1[:, 1, 1]
    orthogonal_vector[:, 1, 1] = vec1[:, 1, 0]
    return orthogonal_vector
```

willingc commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,259 @@
+        res[i * 3 + 1, 1] = (c1 - b1)
+        res[i * 3 + 2, 0] = c1
+        res[i * 3 + 2, 1] = (a1 - c1)
+    return res
```
Perhaps replace `res` with `borders` for better reader comprehension::

```suggestion
    borders = np.empty((len(triangles)*3, 2, 2), dtype=centers.dtype)
    for i, triangle in enumerate(triangles):
        a, b, c = triangle
        a1 = centers[a] + offsets[a]
        b1 = centers[b] + offsets[b]
        c1 = centers[c] + offsets[c]
        borders[i * 3, 0] = a1
        borders[i * 3, 1] = (b1 - a1)
        borders[i * 3 + 1, 0] = b1
        borders[i * 3 + 1, 1] = (c1 - b1)
        borders[i * 3 + 2, 0] = c1
        borders[i * 3 + 2, 1] = (a1 - c1)
    return borders
```

willingc commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,259 @@
+        res[i * 3 + 1, 1] = (c1 - b1)
+        res[i * 3 + 2, 0] = c1
+        res[i * 3 + 2, 1] = (a1 - c1)
+    return res
```
```suggestion
    borders = np.empty((len(triangles)*3, 2, 2), dtype=vertices.dtype)
    for i, triangle in enumerate(triangles):
        a, b, c = triangle
        a1 = vertices[a]
        b1 = vertices[b]
        c1 = vertices[c]
        borders[i * 3, 0] = a1
        borders[i * 3, 1] = (b1 - a1)
        borders[i * 3 + 1, 0] = b1
        borders[i * 3 + 1, 1] = (c1 - b1)
        borders[i * 3 + 2, 0] = c1
        borders[i * 3 + 2, 1] = (a1 - c1)
    return borders
```

willingc commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,259 @@
+
+
+path = np.array([[0,0], [0,1], [1,1], [1,0]]) * 10
+# path = generate_regular_polygon(4, radius=1) * 10
```
Remove comment?

Czaki replied:
```suggestion
```

willingc commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -668,14 +676,15 @@ def triangulate_edge(
             clean_path, closed=closed
         )
 
+    # offsets[2,1] = -0.5
```
Remove or clarify the comment.

willingc reviewed (approved):
@Czaki I've reviewed all of the files. The code looks good to merge as is.

I've made some code maintainability and changes that may help another maintainer or contributor understand this code and reduce cognitive load.

Czaki committed:
Merge branch 'main' into faster_faces

Czaki committed:
Apply suggestions from code review

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
address review code

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,520 @@
+    We use a miter limit instead of maximum miter length
+    for performance reasons.
+    This approach provides an equivalent check.
+    See note in generate_2D_edge_meshes
```
@Czaki sorry that I waited so long after our walkthrough to work on this. I am back at it. 😅 and hope we can get this in ~tomorrow.

In this docstring, I would appreciate definitions and/or links to understand the difference between "miter limit" and "maximum miter length". They sound like exactly the same thing?

Also, "an equivalent check" — what does this mean?

Czaki replied:
Ah. I do not update the name here. 
The `maximum miter length` is the maximum length of miter vector.

The `miter limit` is now `cos_miter_limit` and is the limit for the cosines value. 

It is calculated to be equivalent to checking of vector length, but cheaper to calculate.

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,520 @@
+
+@njit(cache=True)
+def remove_path_duplicates(path: np.ndarray, closed: bool) -> np.ndarray:
+    """Remove duplicates from a path.
```
```suggestion
    """Remove consecutive duplicates from a path.
```

jni commented on napari/layers/shapes/_shapes_models/_polygon_base.py:
```diff
@@ -5,6 +5,15 @@
+except ImportError:
+
+    def remove_path_duplicates(data: np.ndarray, closed: bool):
+        return data
```
@Czaki do we not need to remove duplicates when there is no numba? I can get pretty similar performance with a NumPy version:

```python
def remdup(p):
    diff = np.diff(np.append(p, p[0:1, :], axis=0), axis=0)
    dup = np.all(diff == 0, axis=1)
    indices = np.arange(p.shape[0])
    return p[indices[~dup]]
```

This takes 65µs on a 100 element array with 10 duplicates, vs 50µs on the same array with numba.

jni replied:
(yes I know that we need some more checks but probably not the bottleneck)

Czaki replied:
This code does not work with repeated points on begin/end of a closed path.

jni replied:
I know there are edge cases, this was my quick attempt, but it should be possible to make it work for any data with little loss of performance. My question is whether removing duplicates without numba is completely unnecessary or whether it's useful to come up with a vectorised solution.

jni replied:
Here's the final func I just pushed, which handles both the lasso duplicate points and closed paths:

```python
def _remove_path_duplicates_np(data: np.ndarray, closed: bool):
    # We add the first data point at the end to get the same length bool
    # array as the data, and also to work on closed shapes; the last value
    # in the diff array compares the last and first vertex.
    diff = np.diff(np.append(data, data[0:1, :], axis=0), axis=0)
    dup = np.all(diff == 0, axis=1)
    # if the shape is closed, check whether the first vertex is the same
    # as the last vertex, and count it as a duplicate if so
    if closed and dup[-1]:
        dup[0] = True
    # we allow repeated nodes at the end for the lasso tool, which
    # for an instant needs both the last placed point and the point at the
    # cursor to be the same; if the lasso implementation becomes cleaner,
    # remove this hardcoding
    dup[-2:] = False
    indices = np.arange(data.shape[0])
    return data[indices[~dup]]
```

Czaki committed:
Apply suggestions from code review

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

Czaki committed:
move deduplication to setting data

Czaki committed:
fix

Czaki committed:
fix2

jni committed:
Add docstring to the triangulation example

jni committed:
Improve function docstring in example

jni committed:
Clean up and document generate_order_vectors

jni committed:
Improve docstring of generate_miter_vectors

Czaki committed:
fix typo in variable name

Czaki committed:
fix edge case

jni committed:
Typo fix: mitter -> miter

jni committed:
Improve triangulation docstring and refactor orientation into own function

jni committed:
Disable ruff formatting to get a sensible line break

jni committed:
mitter -> miter

jni committed:
normal vectors -> direction vectors

jni committed:
Complete docstring for generate_miter_vectors

jni committed:
Improve documentation of generate_triangle_borders

jni committed:
Tidy up and annotate shape list

jni committed:
Remove unused mesh variables

jni committed:
Add docstrings for helper data functions

jni committed:
typo fix

jni committed:
minor naming and formatting changes

jni committed:
use dataclass instead of namedtuple

this will enable accessing field names

jni committed:
minor style fix

jni committed:
Automatically add layers based on helper dataclass

jni committed:
Updated names

jni committed:
Add ref points to helper instead of special-casing

jni committed:
Start by selecting shapes layer

jni committed:
Minor cleanups

jni committed:
minor style change

jni committed:
move all imperative code to the end

jni committed:
Move event connection inside adding function

jni committed:
Set zoom and center automatically

jni committed:
protect napari.run under main guard

This will allow adding this example to the gallery if desired

jni committed:
Add docstring to layer adding function

jni committed:
Rename normal_vectors to direction_vectors

"normal vectors" typically mean orthogonal to something. These instead are parallel/tangent to the path. They colud be "normalized" but I think "direction" or "tangent" is the right vocabulary, and I have a slight preference for "direction", as it is less jargony.

jni committed:
Simplify and annotate _calc_output_size

- The _dot operation uses an inline numba function, after benchmarking in IPython (see below). This improves readability

- The repeated code for closed paths is simplified by altering the start and stop values of the loop, since we are using negative indexing anyway.

```
In [28]: @numba.njit(cache=True)
    ...: def subsequent_dots_manual(arr):
    ...:     n = arr.shape[0]
    ...:     output = np.zeros(n - 1)
    ...:     for i in range(n - 1):
    ...:         output[i] = arr[i, 0] * arr[i+1, 0] + arr[i, 1] * arr[i+1, 1
    ...: ]
    ...:     return output
    ...:

In [29]: @numba.njit(cache=True)
    ...: def subsequent_dots_npdot(arr):
    ...:     n = arr.shape[0]
    ...:     output = np.zeros(n - 1)
    ...:     for i in range(n - 1):
    ...:         output[i] = np.dot(arr[i], arr[i+1])
    ...:     return output
    ...:

In [30]: arr = np.random.random((5000000, 2))

In [31]: _ = subsequent_dots_manual(arr)

In [32]: _ = subsequent_dots_npdot(arr)

In [33]: %timeit _ = subsequent_dots_manual(arr)
25 ms ± 1.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [34]: %timeit _ = subsequent_dots_npdot(arr)
565 ms ± 11.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [35]: @numba.njit(cache=True, inline='always')
    ...: def _dot2(arr, i, j):
    ...:     return arr[i, 0] * arr[j, 0] + arr[i, 1] * arr[j, 1]
    ...:

In [36]: @numba.njit(cache=True)
    ...: def subsequent_dots_dot(arr):
    ...:     n = arr.shape[0]
    ...:     output = np.zeros(n - 1)
    ...:     for i in range(n - 1):
    ...:         output[i] = _dot2(arr, i, i+1)
    ...:     return output
    ...:

In [37]: _dot2(arr, 1, 2)
Out[37]: 0.49280610701644534

In [38]: %timeit _ = subsequent_dots_dot(arr)
27.7 ms ± 3.96 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [39]: %timeit _ = subsequent_dots_dot(arr)
24 ms ± 2.96 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [40]: def _dotv(v0, v1):
    ...:     return v0[0] * v1[0] + v0[1] * v1[1]
    ...:

In [41]: @numba.njit(cache=True, inline='always')
    ...: def _dotv(v0, v1):
    ...:     return v0[0] * v1[0] + v0[1] * v1[1]
    ...:

In [42]: @numba.njit(cache=True)
    ...: def subsequent_dots_dotv(arr):
    ...:     n = arr.shape[0]
    ...:     output = np.zeros(n - 1)
    ...:     for i in range(n - 1):
    ...:         output[i] = _dotv(arr[i], arr[i+1])
    ...:     return output
    ...:

In [43]: _ = subsequent_dots_dotv(arr)

In [44]: %timeit _ = subsequent_dots_dotv(arr)
23.7 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [45]: %timeit _ = subsequent_dots_npdot(arr)
565 ms ± 14.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [46]: %timeit _ = np.sum(arr[1:] * arr[:-1], axis=1)
259 ms ± 9.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

jni committed:
Add missing types

jni committed:
Simplify normal vector code branching

jni committed:
Improve normal vector docstring

jni committed:
Rename normal vector func to direction vector

jni committed:
Use dot and cross for cos and sin, add docstrings

jni committed:
Add more explanation around miter point calculation

jni committed:
Simplify clipping of miter scale_factor

jni committed:
Simplify clipping of miter scale_factor further

jni committed:
Optimize sign computation with own func

Rewriting in numba probably avoids some additional checking that the np function must go through.

```
In [1]: import numba

In [2]: import numpy as np
In [3]: arr = np.random.random(5_000_000) - 0.5

In [11]: out = np.empty_like(arr)

In [12]: @numba.njit(cache=True)
    ...: def sign_nb(arr, out):
    ...:     n = len(arr)
    ...:     for i in range(n):
    ...:         out[i] = _sign(arr[i])
    ...:

In [13]: @numba.njit(cache=True)
    ...: def sign_np(arr, out):
    ...:     n = len(arr)
    ...:     for i in range(n):
    ...:         out[i] = np.sign(arr[i])
    ...:

In [14]: sign_np(arr, out)

In [15]: sign_nb(arr, out)

In [16]: %timeit sign_nb(arr, out)
67.9 ms ± 11.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [17]: %timeit sign_np(arr, out)
73.7 ms ± 2.74 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

jni committed:
Simplify and document _set_centers_and_offsets

jni committed:
Add docstring to orthogonal_vector function

jni committed:
simplify _generate_2D_edge_meshes_loop

jni committed:
simplify initial calculation of non-closed miters

jni committed:
Final minor simplification

jni reviewed (approved):
I'm going to approve. My only remaining change is implementing remove_duplicates in pure numpy for when numba isn't installed. But this is otherwise ready to merge and I will merge in a few hours (I've been through the implementation tooth and comb now) to make an rc unless there are any final objections! 🚀

Czaki commented on examples/dev/triangle_edge.py:
```diff
@@ -0,0 +1,379 @@
+def generate_miter_vectors(
+        mesh: tuple[np.ndarray, np.ndarray, np.ndarray]
+        ) -> np.ndarray:
+    """For each point on path, generate the vectors pointing to miter points.
```
It is not fully true. It also works with bevel join. 
So maybe we should update the function name and doc string. 

But I do not know how to name these points that are used to create triangles and lying outside path.

jni replied:
I'm ok calling them "miter points" for now — I certainly do so in my head now after so much time here! 😂 We just happen to use one of the miter points in the bevel join! 😉

jni committed:
One more normal -> direction_vectors rename

jni committed:
Minor comment updates

jni committed:
Add NumPy function for removing path duplicates

jni committed:
Make the remove dup function privately even if numba is available

Czaki commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,601 @@
+        vec_len_arr[i - 1] = np.sqrt(vec_diff[0] ** 2 + vec_diff[1] ** 2)
+        normals[i - 1] = vec_diff / vec_len_arr[i - 1]
+
+    return normals, vec_len_arr * 0.5
```
hm. 
Here I return half of vector length

jni replied:
Ah good, I was confused about the difference between the variable name and the docs (which I did notice said that the miter was limited to half the length) but I couldn't find where it was. Will update the docs!

Czaki commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,601 @@
+            path[i],
+            direction_vectors[i - 1],
+            direction_vectors[i],
+            bevel_limit_array[i - 1],
+            bevel_limit_array[i],
```
And here pass it as vector length

Czaki commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -0,0 +1,601 @@
+        current path vertex (used for bevel join calculation).
+    vec2_len : float
+        The length of the segment between the current path vertex and the
+        next path vertex (used for bevel join calculation).
```
So here we pass half of the vector name when name and description expects full vector length.

Czaki reviewed (commented):
Base on comparing this with c++ implementation (from https://github.com/4DNucleome/PartSegCore-compiled-backend/pull/36) I found one inconsistency between values and documentation

jni committed:
Rename variables and update docstring to reflect half vec len limit

jni committed:
Allow triangulation demo to be run without numba

jni committed:
Use half-length in function name returning vectors
